File: src/California.Creator.Client/Typewriter/StyleAtom.ts


import {CaliforniaProject} from "./CaliforniaProject";
import {StyleAtomType} from "./StyleAtomType";
import {StyleAtomQuantumMapping} from "./StyleAtomQuantumMapping";
import {StyleValue} from "./StyleValue";
import {StyleMoleculeAtomMapping} from "./StyleMoleculeAtomMapping"; 
export class StyleAtom { 
    StyleAtomId: number;
    Name: string;
    StyleAtomType: StyleAtomType;
    AppliedValues: StyleValue[];
    MappedQuantums: StyleAtomQuantumMapping[];
    MappedToMoleculeId: number;
    IsDeletable: boolean;
}

File: src/California.Creator.Service/Models/Core/StyleAtom.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace California.Creator.Service.Models.Core
{
    public class StyleAtom
    {
        public int StyleAtomId { get; set; }
        [PreventDeveloperCodePropagation]
        public int CaliforniaProjectId { get; set; }
        [PreventDeveloperCodePropagation]
        public CaliforniaProject CaliforniaProject { get; set; }

        [Required, StringLength(255)]
        public string Name { get; set; }

        [Required]
        public StyleAtomType? StyleAtomType { get; set; } = Core.StyleAtomType.Generic;

        public List<StyleValue> AppliedValues { get; set; } = new List<StyleValue>(); // TODO constraints to make sure all StyleValues that are mapped to same StyleAtom have different css properties
        public List<StyleAtomQuantumMapping> MappedQuantums { get; set; } = new List<StyleAtomQuantumMapping>();

        public int MappedToMoleculeId { get; set; }
        [JsonIgnore]
        public StyleMoleculeAtomMapping MappedToMolecule { get; set; }

        [NotMapped]
        public bool IsDeletable {
            get
            {
                return AppliedValues.Count == 0;
            }
        }
    }
}
